src/GraphQL/Resolver.hs

```diff
@@ -134,17 +133,24 @@ data Result a = Result [ResolverError] a deriving (Show, Functor, Eq)
 aggregateResults :: [Result Value] -> Result Value
 aggregateResults r = toValue <$> sequenceA r
 
+throwE :: Applicative f => ResolverError -> f (Result Value)
+throwE err = pure (Result [err] GValue.ValueNull)
+
 instance Applicative Result where
   pure v = Result [] v
   (Result e1 f) <*> (Result e2 x) = Result (e1 <> e2) (f x)
 
 ok :: Value -> Result Value
 ok = pure
 
+type AllTypes = Map Name TypeDefinition
 
 class HasResolver m a where
   type Handler m a
-  resolve :: Handler m a -> SelectionSet Value -> m (Result Value)
+  -- XXX: Might be nicer to make AllTypes a ReaderT -- easier to ignore when
+  -- not wanted, and wouldn't have required me to change so many call sites.
+  -- OTOH, with all our type weirdness, I got bored trying to figure it out.
+  resolve :: AllTypes -> Handler m a -> Maybe (SelectionSetByType Value) -> m (Result Value)
```

Comment:
Some reflections. These are mostly unrelated to each other, just a bunch of points loosely on topic.

* What's the `resolve` API for from a users PoV? I genuinely don't know. Knowing that seems the _sine qua non_ in a discussion about usability.
* This is essentially a callback API. It's pretty standard for callbacks to provide arguments that are unused in many cases. We already (before this PR) have many instances that don't use arguments.
* Where callbacks _don't_ provide arguments that are unused, it's because there are two different callback points. This would equate to two different type classes, or two different methods on this one.
* `AllTypes` is something of a stand-in for a value that represents the entire global schema, not just the element that's being resolved. This makes it something more than just an implementation detail.
* If a user _did_ want to write a resolver that had access to the entire schema (as I did), what should they do?
* `ReaderT` is a pretty good compromise here, IMO, since it can be safely ignored by those who don't need it. It also makes it harder to get it wrong by passing a different value into a child resolver (which we alas have to do)
* With all the type magic going on, do you really think an extra argument is the critical usability point?
* If we intend people to provide implementations of `HasResolver`, we should document the class and its methods, and say what the contract of the implementations is.